# Hazem Zaky
# Programming Exercise \#1
---

# Preamble

In [ ]:
# optional code cell when using Google Colab with Google Drive

# remove the docstring comment block below in order to mount Google Drive
'''
# mount Google Drive in Google Colab
from google.colab import drive
drive.mount('/content/drive')

# change directory using the magic command %cd
### replace [MY PATH] below with your own path in Google Drive ###
### %cd /content/drive/My\ Drive/[MY PATH] ###
'''

In [1]:
# import relevant Python libraries

import numpy as np
import pandas as pd
from scipy import stats as sps
from IPython.display import display, Latex

C:\Users\hazem\AppData\Local\Temp\ipykernel_69360\3132124062.py:4: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


# **1. Fetal Health Classification Dataset**

## **Clean Dataset**

### Problem 1.1

In [2]:
# load the clean dataset csv file into a pandas dataframe

fetal_df = pd.read_csv('data/fetal_health_dataset_clean.csv') 

#### (a)

This is a supervised machine learning task because we are given the output of each data value as whether the fetus is normal, suspect, or pathological.

#### (b)

In [7]:
print('Axes: ', fetal_df.axes)
print('\n\n')
print('Dtypes: ', fetal_df.dtypes)

#### (c)

In [11]:
fetal_df.head(10)

baseline value  accelerations  fetal_movement  uterine_contractions  \
0             120          0.000             0.0                 0.000   
1             132          0.006             0.0                 0.006   
2             133          0.003             0.0                 0.008   
3             134          0.003             0.0                 0.008   
4             132          0.007             0.0                 0.008   
5             134          0.001             0.0                 0.010   
6             134          0.001             0.0                 0.013   
7             122          0.000             0.0                 0.000   
8             122          0.000             0.0                 0.002   
9             122          0.000             0.0                 0.003   

   light_decelerations  severe_decelerations  prolongued_decelerations  \
0                0.000                   0.0                     0.000   
1                0.003                   0.0                     0.000   
2                0.003                   0.0                     0.000   
3                0.003                   0.0                     0.000   
4                0.000                   0.0                     0.000   
5                0.009                   0.0                     0.002   
6                0.008                   0.0                     0.003   
7                0.000                   0.0                     0.000   
8                0.000                   0.0                     0.000   
9                0.000                   0.0                     0.000   

   abnormal_short_term_variability  mean_value_of_short_term_variability  \
0                               73                                   0.5   
1                               17                                   2.1   
2                               16                                   2.1   
3                               16                                   2.4   
4                               16                                   2.4   
5                               26                                   5.9   
6                               29                                   6.3   
7                               83                                   0.5   
8                               84                                   0.5   
9                               86                                   0.3   

   percentage_of_time_with_abnormal_long_term_variability  ...  histogram_min  \
0                                                 43       ...             62   
1                                                  0       ...             68   
2                                                  0       ...             68   
3                                                  0       ...             53   
4                                                  0       ...             53   
5                                                  0       ...             50   
6                                                  0       ...             50   
7                                                  6       ...             62   
8                                                  5       ...             62   
9                                                  6       ...             62   

   histogram_max  histogram_number_of_peaks  histogram_number_of_zeroes  \
0            126                          2                           0   
1            198                          6                           1   
2            198                          5                           1   
3            170                         11                           0   
4            170                          9                           0   
5            200                          5                           3   
6            200                          6                           3   
7            130                          0 

#### (d)

In [10]:
fetal_df.shape

(2126, 22)

#### (e)

_[Your answer for 1.1(e) goes here]_

#### (f)

There are 2126 total samples.

#### (g)

21 independent variables.
baseline value
accelerations
fetal_movement
uterine_contractions
light_decelerations
severe_decelerations
prolongued_decelerations
abnormal_short_term_variability
mean_value_of_short_term_variability
percentage_of_time_with_abnormal_long_term_variability
mean_value_of_long_term_variability
histogram_width
histogram_min
histogram_max
histogram_number_of_peaks
histogram_number_of_zeroes
histogram_mode
histogram_mean
histogram_median
histogram_variance
histogram_tendency


#### (h)

1 Dependent variable
fetal_health

#### (i)

n = 2126, p = 21


#### (j)

n = 2126, m = 1

#### (k)

Some of the variables in the dataset are preprocessed variables, and some are raw variables. This is because some of the variables, such as histogram_tendency, histogram_variance, etc. Are based on the general data obtained from the histograms the raw data produced, whereas some variables such as accelerations, fetal_movement, etc. are variables that are obtained directly from the raw data.

#### (l)

In [ ]:
for column in fetal_df.columns:
    unique_values = pd.unique(fetal_df[column])
    print(f"Unique values in '{column}': {unique_values}\n")

histogram_tendency
fetal_health

These are the only categorical variables because the rest of the variables have continuous values that do not have specific bins.


#### (m)

The histogram_tendency variable goes by nominal encoding, where the numbers don't have a specific meaning. The fetal_health variable goes by ordinal encoding because the unique values represent categories of fetal health, with 1 being the healthiest and 3 being the least healthy.

#### (n)

In [28]:
normal = fetal_df[fetal_df['fetal_health'] == 1].count().values[0]
suspect = fetal_df[fetal_df['fetal_health'] == 2].count().values[0]
pathological = fetal_df[fetal_df['fetal_health'] == 3].count().values[0]
print("Normal: ",normal)
print("Suspect: ",suspect)
print("Pathological: ",pathological)


Normal:  1655
Suspect:  295
Pathological:  176


## **Dirty Dataset**

In [29]:
# load the dirty dataset csv file into a pandas dataframe

fetal_dirty_df = pd.read_csv('data/fetal_health_dataset_dirty.csv') 

### Problem 1.2

#### (a)

In [33]:
fetal_dirty_df.isna().sum()

baseline value                                            47
accelerations                                              0
fetal_movement                                             0
uterine_contractions                                       0
light_decelerations                                        0
severe_decelerations                                       0
prolongued_decelerations                                   0
abnormal_short_term_variability                            0
mean_value_of_short_term_variability                       0
percentage_of_time_with_abnormal_long_term_variability    10
mean_value_of_long_term_variability                        0
histogram_width                                            0
histogram_min                                              0
histogram_max                                             24
histogram_number_of_peaks                                  0
histogram_number_of_zeroes                                 0
histogram_mode          

#### (b)

In [36]:
fetal_dirty_df.columns[fetal_dirty_df.isna().sum() != 0].tolist()

['baseline value',
 'percentage_of_time_with_abnormal_long_term_variability',
 'histogram_max',
 'histogram_mode']

#### (c)

In [38]:
fetal_dirty_df[fetal_dirty_df.isna().any(axis=1)].shape[0]

baseline value  accelerations  fetal_movement  uterine_contractions  \
23               NaN          0.000           0.000                 0.000   
43               NaN          0.004           0.012                 0.005   
71             140.0          0.010           0.006                 0.003   
90               NaN          0.002           0.002                 0.002   
93            -148.0          0.000           0.000                 0.000   
...              ...            ...             ...                   ...   
2034             NaN          0.000           0.001                 0.006   
2053          -127.0          0.003           0.003                 0.007   
2059           127.0          0.000           0.001                 0.006   
2066             NaN          0.009           0.003                 0.008   
2109           136.0          0.000           0.001                 0.008   

      light_decelerations  severe_decelerations  prolongued_decelerations  \
23                  0.000                   0.0                     0.000   
43                  0.000                   0.0                     0.000   
71                  0.000                   0.0                     0.000   
90                  0.000                   0.0                     0.000   
93                  0.000                   0.0                     0.000   
...                   ...                   ...                       ...   
2034                0.005                   0.0                     0.002   
2053                0.000                   0.0                     0.000   
2059                0.000                   0.0                     0.000   
2066                0.000                   0.0                     0.000   
2109                0.001                   0.0                     0.000   

      abnormal_short_term_variability  mean_value_of_short_term_variability  \
23                                 80                                   0.5   
43                                 40                                   1.8   
71                                 52                                   1.6   
90                                 64                                   0.4   
93                                 68                                   0.3   
...                               ...                                   ...   
2034                               67                                   3.3   
2053                               65                                   1.2   
2059                               60                                   0.6   
2066                               65                                   1.7   
2109                               67                                   1.9   

      percentage_of_time_with_abnormal_long_term_variability  ...  \
23                                                  0.0       ...   
43                                                  1.0       ...   
71                                                  0.0       ...   
90                                                 29.0       ...   
93                                                 75.0       ...   
...                                                 ...       ...   
2034                                                0.0       ...   
2053                                               46.0       ...   
2059                                               61.0       ...   
2066                                                0.0       ...   
2109                                                0.0       ...   

      histogram_min  histogram_max  histogram_number_of_peaks  \
23              114          130.0                          0   
43               52          194.0                         13   
71               68          172.0                          0   
90              135          163.0                          1   
93              128            NaN                   

### Problem 1.3

In [ ]:
### Your code for 1.3 goes here ###

_[Your justification for 1.3 goes here]_

### Problem 1.4

In [ ]:
### Your code for 1.4 goes here ###

### Problem 1.5

In [ ]:
### Your code for 1.5 goes here ###

### Problem 1.6

In [ ]:
### Your code for 1.6 goes here ###

# **2. Heart Failure Prediction Dataset**

## Problem 2.1

In [ ]:
# load the dataset csv file into a pandas dataframe

### heart_df = pd.read_csv('heart_failure_dataset.csv') ###

### (a)

_[Your answer for 2.1(a) goes here]_

### (b)

In [ ]:
### Your code for 2.1(b) goes here ###

### (c)



In [ ]:
### Your code for 2.1(c) goes here ###

### (d)

In [ ]:
### Your code for 2.1(d) goes here ###

### (e)

_[Your answer for 2.1(e) goes here]_

### (f)

_[Your answer for 2.1(f) goes here]_

### (g)

_[Your answer for 2.1(g) goes here]_

### (h)

_[Your answer for 2.1(h) goes here]_

### (i)

_[Your answer for 2.1(i) goes here]_

### (j)

In [ ]:
### Your code for 2.1(j) goes here ###

_[Your justification for 2.1(j) goes here]_

### (k)

_[Your answer for 2.1(k) goes here]_

### (l)

In [ ]:
### Your code for 2.1(l) goes here ###

### (m)

In [ ]:
### Your code for 2.1(m) goes here ###

### (n)

In [ ]:
### Your code for 2.1(n) goes here ###

## Problem 2.2

In [ ]:
### Your code for 2.2 goes here ###

_[Your justification for 2.2 goes here]_

## Problem 2.3

In [ ]:
### Your code for 2.3 goes here ###

## Problem 2.4

In [ ]:
### Your code for 2.4 goes here ###

## Problem 2.5

In [ ]:
### Your code for 2.5 goes here ###

### (a)

_[Your answer for 2.5(a) goes here]_

### (b)

_[Your answer for 2.5(b) goes here]_

### (c)

_[Your answer for 2.5(c) goes here]_

### (d)

_[Your answer for 2.5(d) goes here]_